multiversx_sdk_core/address.py

```diff
@@ -1,94 +1,106 @@
         return encoded
 
-    def is_smart_contract(self):
-        return self.hex().startswith(SC_HEX_PUBKEY_PREFIX)
+    def get_public_key(self) -> bytes:
+        return self.pubkey
 
-    def get_shard(self) -> int:
-        shard = get_shard_of_pubkey(self.pubkey)
-        return shard
+    def get_hrp(self) -> str:
+        return self.hrp
 
-    def serialize(self) -> bytes:
-        return self.pubkey
+    def is_smart_contract(self):
+        return self.to_hex().startswith(SC_HEX_PUBKEY_PREFIX)
 
-    def __repr__(self):
-        return self.bech32()
+    # this will be removed in v1.0.0; it's here for compatibility reasons with the deprecated transaction builders
+    # the transaction builders will also be removed in v1.0.0
+    def serialize(self) -> bytes:
+        logger.warning("The `serialize()` method is deprecated and will soon be removed")
+        return self.get_public_key()
 
 
-class AddressFactory():
+class AddressFactory:
     def __init__(self, hrp: str = DEFAULT_HRP) -> None:
         self.hrp = hrp
 
-    def create_zero(self) -> Address:
-        return Address(bytearray(32), self.hrp)
-
     def create_from_bech32(self, value: str) -> Address:
         hrp, pubkey = _decode_bech32(value)
         if hrp != self.hrp:
             raise ErrBadAddress(value)
 
         return Address(pubkey, hrp)
 
-    def create_from_pubkey(self, pubkey: bytes) -> Address:
+    def create_from_public_key(self, pubkey: bytes) -> Address:
         return Address(pubkey, self.hrp)
 
     def create_from_hex(self, value: str) -> Address:
-        return Address.from_hex(value, self.hrp)
+        return Address.new_from_hex(value, self.hrp)
 
 
-class AddressConverter():
-    def __init__(self, hrp: str = DEFAULT_HRP) -> None:
-        self.hrp = hrp
+class AddressComputer:
+    def __init__(self) -> None:
+        pass
 
-    def bech32_to_pubkey(self, value: str) -> bytes:
-        hrp, pubkey = _decode_bech32(value)
-        if hrp != self.hrp:
-            raise ErrBadAddress(value)
-
-        return pubkey
+    def compute_contract_address(self, deployer: IAddress, deployment_nonce: int) -> Address:
+        """
+        8 bytes of zero + 2 bytes for VM type + 20 bytes of hash(owner) + 2 bytes of shard(owner)
+        """
+        deployer_pubkey = deployer.get_public_key()
+        nonce_bytes = deployment_nonce.to_bytes(8, byteorder="little")
+        bytes_to_hash = deployer_pubkey + nonce_bytes
+        contract_pubkey = keccak.new(digest_bits=256).update(bytes_to_hash).digest()
+        contract_pubkey = bytes([0] * 8) + bytes([5, 0]) + contract_pubkey[10:30] + deployer_pubkey[30:]
+        return Address(contract_pubkey, deployer.get_hrp())
 
-    def pubkey_to_bech32(self, pubkey: bytes) -> str:
-        address = Address(pubkey, self.hrp)
-        return address.bech32()
+    def get_shard_of_address(self, address: IAddress) -> int:
+        return get_shard_of_pubkey(address.get_public_key())
```

Comment:
in the specs it's specified that number of shards will be passes via the constructor, should it be done here?

Reply:
Passed it in via the constructor and is now passed to the `get_shard_of_pubkey()` function.